src/reducers/fileTree.tsx

```diff
@@ -191,6 +190,100 @@ export const getRelativePath = ({
   return pathList[newIndex];
 };
 
+const findRelativeMessageUid = (
+  currentPath: string,
+  messageMap: LinterMessageMap,
+  pathList: string[],
+  position: RelativePathPosition,
+): LinterMessage['uid'] | null => {
+  let path = currentPath;
+  for (let i = 0; i < pathList.length; i++) {
+    const nextPath = getRelativePath({
+      currentPath: path,
+      pathList,
+      position,
+    });
+    const nextMessages = messageMap[nextPath];
+    if (nextMessages) {
+      const nextMessagesForPath = getMessagesForPath(nextMessages);
+      if (position === RelativePathPosition.previous) {
+        const lastMessageIndex = nextMessagesForPath.length - 1;
+        if (lastMessageIndex >= 0) {
+          return nextMessagesForPath[lastMessageIndex].uid;
+        }
+      } else if (nextMessagesForPath.length) {
+        return nextMessagesForPath[0].uid;
+      }
+    }
+    path = nextPath;
+  }
+  // If we got here then we never found another message, which would be odd,
+  // because we should have at the very least gotten back to our original
+  // message.
+  // We should probably log this.
+  return null;
+};
+
+export type GetRelativeMessageUidParams = {
+  currentMessageUid: LinterMessage['uid'] | void;
+  currentPath: string;
+  locateCurrentMessage?: boolean;
+  messageMap: LinterMessageMap;
+  pathList: string[];
+  position: RelativePathPosition;
+};
+
+export const getRelativeMessageUid = ({
+  currentMessageUid,
+  currentPath,
+  messageMap,
+  pathList,
+  position,
+}: GetRelativeMessageUidParams): string | null => {
+  if (!currentMessageUid) {
```

Comment:
The edge case of not looking for a relative message is simpler but it's not much different than looking for a relative message. You could combine the logic so that there aren't so many code paths -- the fewer diverging code paths the better.

After improving test coverage, you could try combining them like this:

```diff
diff --git a/src/reducers/fileTree.tsx b/src/reducers/fileTree.tsx
index c956ffd..83093f4 100644
--- a/src/reducers/fileTree.tsx
+++ b/src/reducers/fileTree.tsx
@@ -240,25 +240,9 @@ export const getRelativeMessageUid = ({
   pathList,
   position,
 }: GetRelativeMessageUidParams): string | null => {
-  if (!currentMessageUid) {
-    // There is no current message, so get the first one.
-    if (!Object.keys(messageMap).length) {
-      return null;
-    }
-    for (const path of pathList) {
-      const messages = messageMap[path];
-      if (messages) {
-        const messagesForPath = getMessagesForPath(messages);
-        if (messagesForPath.length) {
-          return messagesForPath[0].uid;
-        }
-      }
-    }
-    // No messages found in the map (which shouldn't be possible because of the
-    // length check, so throw an error.
-    throw new Error(
-      'The messageMap is not empty, but we were unable to find any messages',
-    );
+  if (!Object.keys(messageMap).length) {
+    // There are no messages at all.
+    return null;
   }
 
   const messages = messageMap[currentPath];
@@ -269,14 +253,22 @@ export const getRelativeMessageUid = ({
   }
 
   const messagesForPath = getMessagesForPath(messages);
-  const currentMessageIndex = messagesForPath.findIndex(
-    (message) => message.uid === currentMessageUid,
-  );
 
-  const newIndex =
-    position === RelativePathPosition.previous
-      ? currentMessageIndex - 1
-      : currentMessageIndex + 1;
+  let newIndex;
+  if (!currentMessageUid) {
+    // Since we aren't looking a message relative to an existing one,
+    // just get the first message.
+    newIndex = 0;
+  } else {
+    const currentMessageIndex = messagesForPath.findIndex(
+      (message) => message.uid === currentMessageUid,
+    );
+
+    newIndex =
+      position === RelativePathPosition.previous
+        ? currentMessageIndex - 1
+        : currentMessageIndex + 1;
+  }
 
   if (newIndex >= 0 && newIndex < messagesForPath.length) {
     return messagesForPath[newIndex].uid;
```

Reply:
Yes, that's much better. Thanks!